### ISSUE TITLE
Links may be opened by wrong apps

### ISSUE THREAD

**Issue by tsujan** [lxqt/lxqt#1111]
LXQt is seen as a "generic" DE by `xdg-mime`; therefore, file types are determined by the function `info_generic()` inside `xdg-mime`. This function searches for the Perl script `mimetype` (provided by `libfile-mimeinfo-perl` in Debian based systems) and if not found, uses `file` to know the file type. 

This means that if the Perl script `mimetype` doesn't exist, file types will be determined only by consulting `/usr/share/mime/magic` and without considering `/usr/share/mime/packages`, which might result in opening links with wrong apps.

For example, if a mimetype is derived from xml, it'll always be opened by Firefox or any app that deals with xml, while it should be opened by its corresponding app.

`gvfs-open` and `exo-open` are independent from `libfile-mimeinfo-perl` but they aren't used by LXQt.

In short, LXQt needs `libfile-mimeinfo-perl` (or `perl-file-mimeinfo` in Arch) for `xdg-open` to work correctly. This dependency should be added to lxqt-session, IMO.

**Comment by paulolieuthier** [comment]
This should be a recommended dependency, for packagers to add. I'm not sure if we should enforce it.

**Comment by agaida** [comment]
This is a packaging thing - please place it in big fat letters in the README.md - maybe we should list dependencies and recommends in the README for all packages.

**Comment by tsujan** [comment]
@paulolieuthier @agaida 
We can't expect our users to know about such technicalities just in order to open links correctly. We can't expect them to read `README.md` either. They might wonder why links aren't opened correctly under LXQt, while they're opened right under KDE, Gnome, XFCE, Enlightenment....

On the other hand, there's no enforcing in this: `libfile-mimeinfo-perl` is a little package that's installed on any Debian-based system -- but not on Arch-based systems!

**Comment by tsujan** [comment]
Actually, KDE, Gnome, XFCE,... have their own ways of dealing with this issue. LXQt has none. So, either LXQt should have its own way or it should depend on `libfile-mimeinfo-perl`.

**Comment by agaida** [comment]
@tsujan - i bet that 90-99% of our users use distribution packages - and for users that compile  LXQt it's a must to read the READMEs. And you are not 100% right about the assumption on debian systems and derivatives:
```
$ LANG=C aptitude why libfile-mimeinfo-perl 
i   xdg-utils Recommends libfile-mimeinfo-perl
```
So, if one install without recommends - some derivative do that - libfile-mimeinfo-perl will not be installed automagically.

**Comment by agaida** [comment]
More verbose:
```
$ LANG=C apt rdepends libfile-mimeinfo-perl 
libfile-mimeinfo-perl
Reverse Depends:
  Depends: auto-multiple-choice-common
  Recommends: xdg-utils
  Suggests: libio-all-perl
  Depends: libebook-tools-perl
  Depends: ledgersmb
  Depends: ikiwiki-hosting-web
  Suggests: ikiwiki
  Depends: findimagedupes
  Depends: auto-multiple-choice-common
  Depends: ledgersmb
  Recommends: xdg-utils
  Suggests: libio-all-perl
  Depends: libebook-tools-perl
  Depends: ledgersmb
  Depends: ikiwiki-hosting-web
  Suggests: ikiwiki
  Depends: findimagedupes
```

**Comment by tsujan** [comment]
@agaida I understand your reasoning. However, this is a logical issue too. As I said above, LXQt really lacks a method of its own for dealing with mimetypes when links are opened. Maybe such a mechanism should be added to LXQt but until then, `libfile-mimeinfo-perl` is logically not  just a recommendation but a necessity, IMO.

**Comment by agaida** [comment]
For debian systems it's easy - we can enforce it: 
```
Package: libqt5xdg1
Architecture: any
Multi-Arch: same
Depends: ${shlibs:Depends},
         ${misc:Depends}.
         libfile-miminfo-perl
```
if should be there is not enough - and it's not only recommended to have this lib installed i would add this as a dependency.

**Comment by tsujan** [comment]
@agaida That would be nice! And I didn't know that not all Debian-based systems take `libfile-mimeinfo-perl` for granted.

**Comment by agaida** [comment]
Dependencies and Recommends are ever a hard struggle in Debian - it's about the policy - if a package work without another package, but this other package add functionality - it's a Recommend. And we can now discuss if choosing the right app to open a file is base functionality - ok. in debian recommends will always be installed - unless one has very good reasons to turn this off. So if we really want to have this package installed, Recommend is not enough, we must depend on it.

**Comment by tsujan** [comment]
@agaida On Debian, I always disabled installtion of Recommends;) Anyway, what I meant above was about the installable ISO images of Debian and its derivatives: I thought all of them include `libfile-mimeinfo-perl`, while that's not the case about Manjaro, let alone about Arch, whose installation is manual.

**Comment by agaida** [comment]
And really nice - as i don't know it:
```
~$ grep -r Recommends /etc/apt
/etc/apt/apt.conf.d/99synaptic:APT::Install-Recommends "false";
/etc/apt/apt.conf.d/80siduction:// Recommends are as of now still abused in many packages
/etc/apt/apt.conf.d/80siduction:APT::Install-Recommends "0";
```
so a system with installed synaptic behave other than a system without :) - i don't know if 99synaptic is still valid, but there was some discussions about

**Comment by agaida** [comment]
@tsujan - but the ISO images without X will not have that file - as it is recommend by xdg-utils - catch 22. And if one debootstrap a system (many people do this i guess) it will not be installed in some cases.

**Comment by tsujan** [comment]
@agaida OK! So we both agree that `libfile-mimeinfo-perl` should be a dependency -- even on Debian!

**Comment by agaida** [comment]
right - i wil change this right now.

**Comment by stefonarch** [comment]
It's more than a year that I use a symlink (on arch) which replaces xdg-open by kde-open because I had weird behaviours which I could not solve using file associations (qupzilla opened pdf's in gimp and others), this may be the explanation,  perl-file-mimeinfo was not installed.

**Comment by agaida** [comment]
@tsujan: thanks again, uploaded - https://anonscm.debian.org/git/pkg-lxqt/libqtxdg.git/?h=debian%2Fsid

**Comment by tsujan** [comment]
@agaida Great!

@stefonarch Today, I was going to replace `xdg-open` with `gvfs-open` on my newly installed Manjaro for the same reason but remembered that I had no problem on Debian. After playing with `gvfs-open` and `gvfs-mime`, changing some of their lines, adding `echo ...` to some places, etc. I found that a little Perl script was missing ;)
